### PR TITLE
Enable multi-platform image building and add integration test

### LINKED ISSUE #2523: Push multi-platform images together in one go (i.e., push manifest list instead of manifest)
docker buildx allows building for multiple archs (arm64, amd64) in one go. It would be nice if jib could do the same.  

See https://github.com/docker/buildx#building-multi-platform-images for more info.

### PR DESCRIPTION
**Pull request by louismurerwa**
Closes #2523.

Removes the guard for the multi-platform image building feature.

Adds an integration test.

### CODE REVIEW

File: jib-maven-plugin/src/test/resources/maven/projects/simple/pom-multiplatform-build.xml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  <dependencies>
+    <dependency>
+      <groupId>com.test</groupId>
+      <artifactId>dependency</artifactId>
+      <version>1.0.0</version>
+      <scope>system</scope>
+      <systemPath>${project.basedir}/libs/dependency-1.0.0.jar</systemPath>
+    </dependency>
+  </dependencies>
+
```
Comment (chanseokoh) [inline jib-maven-plugin/src/test/resources/maven/projects/simple/pom-multiplatform-build.xml:25]: See if you can remove this.

File: jib-maven-plugin/src/test/resources/maven/projects/simple/pom-multiplatform-build.xml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            <args>An argument.</args>
+            <ports>
+              <port>1000/tcp</port>
+              <port>2000-2003/udp</port>
+            </ports>
+            <labels>
+              <key1>value1</key1>
+              <key2>value2</key2>
+            </labels>
+          </container>
```
Comment (chanseokoh) [inline jib-maven-plugin/src/test/resources/maven/projects/simple/pom-multiplatform-build.xml:72]: Probably we can remove the whole `<container>`.

File: jib-maven-plugin/src/test/resources/maven/projects/simple/pom-multiplatform-build.xml [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	    <platform>
+	       <architecture>amd64</architecture>
+	       <os>linux</os>
+	    </platform>
+	   </platforms>
+          </from> 
+          <to>
+            <image>${_TARGET_IMAGE}</image>
+             <tags>
+              <tag>latest</tag>
```
Comment (chanseokoh) [inline jib-maven-plugin/src/test/resources/maven/projects/simple/pom-multiplatform-build.xml:59]: Oh, how about also testing multiple tags?

```suggestion
              <tag>latest</tag>
              <tag>another</tag>
```

You can then make two `pullManifest()` calls and compare them they are identical.

File: jib-maven-plugin/src/test/resources/maven/projects/simple/pom-multiplatform-build.xml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        </configuration>
+      </plugin>
+
+      <plugin>
+        <groupId>com.google.cloud.tools</groupId>
+        <artifactId>jib-maven-plugin</artifactId>
+        <version>${jib-maven-plugin.version}</version>
+        <configuration>
+           <from>
+        <image>busybox</image>
```
Comment (chanseokoh) [inline jib-maven-plugin/src/test/resources/maven/projects/simple/pom-multiplatform-build.xml:44]: `busybox@sha256:4f47c01fa91355af2865ac10fef5bf6ec9c7f42ad2321377c21e844427972977`

This is the digest of the current manifest list I see today.

**Comment by chanseokoh** [comment]
This one's ready. This finally enables the multi-platform image building (by removing the guard) and adds an integration test. I'll add CHANGELOG soon, but other than that, ready for review.

File: jib-maven-plugin/src/integration-test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java [hunk 5]
```diff
@@ -686,6 +695,85 @@ public void testExecute_springBootPackaged()
...
+        RegistryClient.factory(EventHandlers.NONE, "localhost:5000", "multiplatform", httpClient)
+            .setCredential(Credential.from("testuser", "testpassword"))
+            .newRegistryClient();
+    registryClient.configureBasicAuth();
+
+    // manifest list by tag ":latest"
+    ManifestTemplate manifestListTemplate = registryClient.pullManifest("latest").getManifest();
+    MatcherAssert.assertThat(
+        manifestListTemplate, CoreMatchers.instanceOf(V22ManifestListTemplate.class));
+    V22ManifestListTemplate manifestList = (V22ManifestListTemplate) manifestListTemplate;
```
Comment (mpeddada1) [inline jib-maven-plugin/src/integration-test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java:728]: Could we move this to be below line 725? To keep all the assert statements together.

File: jib-maven-plugin/src/integration-test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java [hunk 6]
```diff
@@ -686,6 +695,85 @@ public void testExecute_springBootPackaged()
...
+        Arrays.asList("sha256:fee2655e19e5138150606c99cfc16fcbf502d72b0f3b9ccf3a8f4509c47e46d9"),
+        manifestList2.getDigestsForPlatform("arm64", "linux"));
+    Assert.assertEquals(
+        Arrays.asList("sha256:f3f4a91c68bcafea351280085d17e25fa598f5644c8b5e31e6133eddfc35e7ff"),
+        manifestList2.getDigestsForPlatform("amd64", "linux"));
+
+    // arm64/linux manifest
+    ManifestAndDigest<ManifestTemplate> manifestAndDigest1 =
+        registryClient.pullManifest(
+            "sha256:fee2655e19e5138150606c99cfc16fcbf502d72b0f3b9ccf3a8f4509c47e46d9");
```
Comment (mpeddada1) [inline jib-maven-plugin/src/integration-test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java:751]: maybe we can store this in a variable?

File: jib-maven-plugin/src/integration-test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java [hunk 7]
```diff
@@ -686,6 +695,85 @@ public void testExecute_springBootPackaged()
...
+    Assert.assertEquals(
+        "sha256:fee2655e19e5138150606c99cfc16fcbf502d72b0f3b9ccf3a8f4509c47e46d9",
+        manifestAndDigest1.getDigest().toString());
+    ManifestTemplate manifestTemplate1 = manifestAndDigest1.getManifest();
```
Comment (mpeddada1) [inline jib-maven-plugin/src/integration-test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java:755]: Can we move line 755 to be below line 749? To make the assert blocks be separate from the initialization code.

File: jib-maven-plugin/src/integration-test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java [hunk 8]
```diff
@@ -686,6 +695,85 @@ public void testExecute_springBootPackaged()
...
+    MatcherAssert.assertThat(manifestTemplate1, CoreMatchers.instanceOf(V22ManifestTemplate.class));
+    V22ManifestTemplate manifest1 = (V22ManifestTemplate) manifestTemplate1;
+    Assert.assertEquals(
+        "sha256:cecb4d0f179207a1c7f2ee33819d4fb70bbb9d98eebe78dfe1b439896925dc27",
+        manifest1.getContainerConfiguration().getDigest().toString());
+
+    // amd64/linux manifest
+    ManifestAndDigest<ManifestTemplate> manifestAndDigest2 =
+        registryClient.pullManifest(
+            "sha256:f3f4a91c68bcafea351280085d17e25fa598f5644c8b5e31e6133eddfc35e7ff");
```
Comment (mpeddada1) [inline jib-maven-plugin/src/integration-test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java:765]: maybe have a variable for this as well.

File: jib-maven-plugin/src/integration-test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java [hunk 9]
```diff
@@ -686,6 +695,85 @@ public void testExecute_springBootPackaged()
...
+    Assert.assertEquals(
+        "sha256:f3f4a91c68bcafea351280085d17e25fa598f5644c8b5e31e6133eddfc35e7ff",
+        manifestAndDigest2.getDigest().toString());
+    ManifestTemplate manifestTemplate2 = manifestAndDigest2.getManifest();
```
Comment (mpeddada1) [inline jib-maven-plugin/src/integration-test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java:769]: same here, we could move this to be below line 765?

**Review by mpeddada1 — COMMENTED** [review]
Very exciting!

File: jib-maven-plugin/src/integration-test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java (hunk 6 above)
Comment (chanseokoh) [reply jib-maven-plugin/src/integration-test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java:751]: Not always, but we usually don't create a variable for constants in tests.

File: jib-maven-plugin/src/integration-test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java [hunk 10]
```diff
@@ -686,6 +695,85 @@ public void testExecute_springBootPackaged()
...
+    // arm64/linux manifest
+    ManifestAndDigest<ManifestTemplate> manifestAndDigest1 =
+        registryClient.pullManifest(
+            "sha256:fee2655e19e5138150606c99cfc16fcbf502d72b0f3b9ccf3a8f4509c47e46d9");
+    ManifestTemplate manifestTemplate1 = manifestAndDigest1.getManifest();
+    Assert.assertEquals(
+        "sha256:fee2655e19e5138150606c99cfc16fcbf502d72b0f3b9ccf3a8f4509c47e46d9",
+        manifestAndDigest1.getDigest().toString());
+    MatcherAssert.assertThat(manifestTemplate1, CoreMatchers.instanceOf(V22ManifestTemplate.class));
+    V22ManifestTemplate manifest1 = (V22ManifestTemplate) manifestTemplate1;
```
Comment (mpeddada1) [inline jib-maven-plugin/src/integration-test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java:757]: Sorry, can we move this up to line 753 as well. Similar to what we do in line 737-> line 740. Same below.

Comment (chanseokoh) [reply jib-maven-plugin/src/integration-test/java/com/google/cloud/tools/jib/maven/BuildImageMojoIntegrationTest.java:757]: The cast should be below the assert above. The assertion is to ensure that the template is of type `V22ManifestTemplate` before attempting casting.

**Review by mpeddada1 — APPROVED** [review]

**Comment by chanseokoh** [comment]
Hmm... the Kokoro Mac integration test failure is interesting. I saw that once locally, which I thought was only because my branch was not up-to-date, but I think it's not. It's the order of platforms in the manifest list that matters here. I didn't see any places of potential non-deterministic behavior. This is strange.

**Pull request merged**